### PR TITLE
ref(ci): Use a single CI workflow for tests

### LINKED ISSUE #7816: devops: Build CI Docker runs twice for every PR
### Describe the issue or request

On every PR, we run the same Docker build twice, with exactly the same settings:

https://github.com/ZcashFoundation/zebra/blob/aad883b68ae1fb7a98aa2868bc1bd4a34954bd13/.github/workflows/ci-unit-tests-docker.yml#L84-L86

https://github.com/ZcashFoundation/zebra/blob/aad883b68ae1fb7a98aa2868bc1bd4a34954bd13/.github/workflows/ci-integration-tests-gcp.yml#L126-L128

### Expected Behavior

This seems redundant, we only need to build one image.

### Current Behavior

The same image is built twice.

### Possible Solution

Make the jobs in one workflow wait on the build job in the other workflow, or re-combine the workflows into a single workflow.

### Additional Information/Context

This costs us extra build machine time, and extra storage costs.

### Is this happening on PRs?

Yes

### Is this happening on the main branch?

Not really, de-duplicating the `main` merge and release builds is more complicated, and there is only one duplicate every few weeks

### PR DESCRIPTION
**Pull request by gustavovalverde**
## Motivation

This is an initial implementation to solve the re-building of our Docker image, which is being built multiple times by our Github Actions, causing elevated costs across multiple services.

Fixes #7816

#### Check before marking the PR as ready for review:
  - [ ] Will the PR name make sense to users?
  - [ ] Does the PR have a priority label?
  - [ ] Have you added or updated tests?
  - [ ] Is the documentation up to date?

##### For significant changes:
  - [ ] Is there a summary in the CHANGELOG?
  - [ ] Can these changes be split into multiple PRs?

_If a checkbox isn't relevant to the PR, mark it as done._

### Complex Code or Requirements

- This required creating an extra workflow `ci-tests.yml` with its respectives `.patch.yml` and `.patch-external.yml` to initialize the unit tests and integration tests from a single place, effectively using a single Docker image build process.
- This must be admin merged, and all open PRs must be rebased afterwards.

## Solution

- Create a `Run tests` [starter workflow](https://docs.github.com/en/actions/using-workflows/reusing-workflows#reusable-workflows-and-starter-workflows)
- Convert the `Integration Tests on GCP` and `Docker Unit Tests` to [callable workflows](https://docs.github.com/en/actions/using-workflows/reusing-workflows#reusable-workflows-and-starter-workflows)
- Merge both of the `Integration Tests on GCP` and `Docker Unit Tests` patches files into a single patch file for `Run tests`

### Testing

- Tests should run on this PR successfully

### Reviewer Checklist

Check before approving the PR:
  - [ ] Does the PR scope match the ticket?
  - [ ] Are there enough tests to make sure it works? Do the tests cover the PR motivation?
  - [ ] Are all the PR blockers dealt with?
        PR blockers can be dealt with in new tickets or PRs.

_And check the PR Author checklist is complete._

## Follow Up Work

This must be followed by https://github.com/ZcashFoundation/zebra/pull/8374 to effectively reduce our costs.

### CODE REVIEW

File: .github/workflows/ci-tests.yml [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+      # dependencies
+      - "**/Cargo.toml"
+      - "**/Cargo.lock"
+      # configuration files
+      - ".cargo/config.toml"
+      - "**/clippy.toml"
+      # workflow definitions
+      - "docker/**"
+      - ".dockerignore"
+      - ".github/workflows/ci-tests.yml"
```
Comment (upbqdn) [inline .github/workflows/ci-tests.yml:69]: Should we also add the following workflows?

```suggestion
      - ".github/workflows/ci-tests.yml"
      - ".github/workflows/sub-ci-unit-tests-docker.yml"
      - ".github/workflows/sub-ci-integration-tests-gcp.yml"
```

File: .github/workflows/ci-tests.yml [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+        default: false
+        description: "Force tests to always create a cached state disk, if they already create disks"
+      no_cache:
+        description: "Disable the Docker cache for this build"
+        required: false
+        type: boolean
+        default: false
+
+  pull_request:
+    # Skip PRs where Rust code and dependencies aren't modified.
```
Comment (upbqdn) [inline .github/workflows/ci-tests.yml]: ###### Optional Suggestion

```suggestion
    # Run only on PRs that modify Rust code or dependencies.
```

File: .github/workflows/ci-tests.yml [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+      # workflow definitions
+      - "docker/**"
+      - ".dockerignore"
+      - ".github/workflows/ci-tests.yml"
+      - ".github/workflows/sub-deploy-integration-tests-gcp.yml"
+      - ".github/workflows/sub-find-cached-disks.yml"
+      - ".github/workflows/sub-build-docker-image.yml"
+
+  push:
+    # Skip main branch updates where Rust code and dependencies aren't modified.
```
Comment (upbqdn) [inline .github/workflows/ci-tests.yml]: ###### Optional Suggestion

```suggestion
    # Run only on main branch updates that modify Rust code or dependencies.
```

File: .github/workflows/ci-tests.patch-external.yml [hunk 4]
```diff
@@ -1,115 +1,139 @@
...
+#! IMPORTANT
+#!
+#! The job names in `sub-ci-unit-tests-docker.yml`, `sub-ci-integration-tests-gcp.yml`,
+#! `ci-tests.patch.yml` and `ci-tests.patch-external.yml` must be kept in sync.
 jobs:
-  # We don't patch the testnet job, because testnet isn't required to merge (it's too unstable)
-  get-available-disks:
-    name: Check if cached state disks exist for Mainnet / Check if cached state disks exist
+  build:
+    name: Build CI Docker / Build images
```
Comment (upbqdn) [inline .github/workflows/ci-tests.patch-external.yml:16]: Should we change the name to match the job name in `ci-tests.yml`?

```suggestion
    name: Build CI Docker / Build CI image
```

**Review by upbqdn — COMMENTED** [review]
This PR removes workflow patches for unit tests, but it doesn't do so for integration tests. Why is that?

File: .github/workflows/ci-tests.patch.yml [hunk 5]
```diff
@@ -1,45 +1,98 @@
...
+#! IMPORTANT
+#!
+#! The job names in `sub-ci-unit-tests-docker.yml`, `sub-ci-integration-tests-gcp.yml`,
+#! `ci-tests.patch.yml` and `ci-tests.patch-external.yml` must be kept in sync.
 jobs:
-  # We don't patch the testnet job, because testnet isn't required to merge (it's too unstable)
-  get-available-disks:
-    name: Check if cached state disks exist for Mainnet / Check if cached state disks exist
+  build:
+    name: Build CI Docker / Build images
```
Comment (upbqdn) [inline .github/workflows/ci-tests.patch.yml:37]: Related to a previous comment.

```suggestion
    name: Build CI Docker / Build CI image
```

**Comment by gustavovalverde** [comment]
> This PR removes workflow patches for unit tests, but it doesn't do so for integration tests. Why is that?

The changelog is confusing, but I removed the unit tests patches, and renamed and reused the same patch files for integration tests, those are the new `ci-tests.patch.yml` and `ci-tests.patch-external.yml`. This was mainly to reduce the diff.

File: .github/workflows/ci-tests.patch-external.yml [hunk 6]
```diff
@@ -1,115 +1,139 @@
...
+
+  test-zebra-conf-path:
+    name: Test CI custom Docker config file / Test custom-conf in Docker
+    runs-on: ubuntu-latest
+    steps:
+      - run: 'echo "Skipping job on fork"'
+
+
+  ####
+  ## The following jobs are related to sub-ci-intgration-tests-gcp.yml
```
Comment (upbqdn) [inline .github/workflows/ci-tests.patch-external.yml]:
```suggestion
  ## The following jobs are related to sub-ci-integration-tests-gcp.yml
```

File: .github/workflows/ci-tests.patch.yml [hunk 7]
```diff
@@ -1,45 +1,98 @@
...
+
+  test-zebra-conf-path:
+    name: Test CI custom Docker config file / Test custom-conf in Docker
+    runs-on: ubuntu-latest
+    steps:
+      - run: 'echo "No build required"'
+
+
+  ####
+  ## The following jobs are related to sub-ci-intgration-tests-gcp.yml
```
Comment (upbqdn) [inline .github/workflows/ci-tests.patch.yml]:
```suggestion
  ## The following jobs are related to sub-ci-integration-tests-gcp.yml
```

**Review by upbqdn — DISMISSED** [review]

**Comment by gustavovalverde** [comment]
@upbqdn after merging your last suggestions, this will require re-approval. Thanks!

**Review by upbqdn — DISMISSED** [review]

**Comment by gustavovalverde** [comment]
@upbqdn Oops! I had to revert one of the suggestions as I had forgotten the reasoning behind it.

The job is called `Build images` as the specific workflow that's being (`./.github/workflows/sub-build-docker-image.yml`) called is used for **CI**, **CD** and **Release** images.

Also, changing the job name here would then require 2 merge rules instead of 1:

Now:
- `Build CI Docker / Build images` and `Build CD Docker / Build images` (1 rule)
After:
- `Build CI Docker / Build CI image` (1 rule)
- `Build CD Docker / Build CD image` (1 rule)

This would also be a breaking change, requiring other PRs to be updated. I'd prefer to avoid that in this refactor.

File: .github/workflows/ci-tests.patch.yml [hunk 8]
```diff
@@ -1,45 +1,98 @@
...
+#! IMPORTANT
+#!
+#! The job names in `sub-ci-unit-tests-docker.yml`, `sub-ci-integration-tests-gcp.yml`,
+#! `ci-tests.patch.yml` and `ci-tests.patch-external.yml` must be kept in sync.
 jobs:
-  # We don't patch the testnet job, because testnet isn't required to merge (it's too unstable)
-  get-available-disks:
-    name: Check if cached state disks exist for Mainnet / Check if cached state disks exist
+  build:
+    name: Build CI Docker / Build CI image
```
Comment (upbqdn) [inline .github/workflows/ci-tests.patch.yml]: I guess we should change this one then as well.

```suggestion
    name: Build CI Docker / Build images
```

**Review by upbqdn — DISMISSED** [review]

**Review by upbqdn — APPROVED** [review]

**Pull request merged**